### PR TITLE
(WIP) CRM-20263: Improve support for Accrual Method bookkeeping 12

### PR DESCRIPTION
**Pull request by pradpnayak**
* [CRM-20263: Accrual accounting entries are not created correctly for fixed membership](https://issues.civicrm.org/jira/browse/CRM-20263)

### CODE REVIEW

**Comment by Edzelopez** [comment]
Rebased PR

**Comment by monishdeb** [comment]
Jenkin test this please

**Comment by monishdeb** [comment]
@pradpnayak can you please fix these related test failures https://test.civicrm.org/job/CiviCRM-Core-PR/14362/

**Comment by colemanw** [comment]
@civicrm-builder retest this please

**Comment by eileenmcnaughton** [comment]
Just a quick not that the JIRA needs more detail about what is being fixed here

**Comment by monishdeb** [comment]
Fixed style warnings and add related [test build failures](https://test.civicrm.org/job/CiviCRM-Core-PR/14873/)

**Comment by monishdeb** [comment]
@pradpnayak I have fix one of the two related unit test failure. Can you fix the other https://test.civicrm.org/job/CiviCRM-Core-PR/15204/testReport/junit/(root)/api_v3_ReportTemplateTest/testDeferredRevenueReport/ and also update the CRM-20263 description, with steps to replicate the issue ?

**Comment by pradpnayak** [comment]
sure i will update the ticket

**Comment by eileenmcnaughton** [comment]
Just a quick note - can you just look at the PR title & commit messages when you have this all working - so it's clearer what it is fixing

File: CRM/Contribute/BAO/Query.php [hunk 1]
```diff
@@ -854,6 +854,7 @@ public static function defaultReturnProperties($mode, $includeCustomFields = TRU
         // calling
         'contribution_product_id' => 1,
         //function
+        'revenue_recognition_date' => 1,
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Query.php]: Can we avoid calling kittens? ie. set the return properties where required

File: CRM/Price/BAO/LineItem.php [hunk 2]
```diff
@@ -477,6 +477,7 @@ public static function processPriceSet($entityId, $lineItem, $contributionDetail
         }
       }
     }
+    // create deferred revenue transaction
```
Comment (eileenmcnaughton) [inline CRM/Price/BAO/LineItem.php]: This comment doesn't really tell use anything does it? We are calling createDeferredTrxn

**Comment by eileenmcnaughton** [comment]
Overall it looks like this PR would not have fallout & has tests. I've made some minor comments.

On a bigger stylistic level I think the code could be restructured to make the interaction  more disciplined. Where we have 'add on functionality' in core I think we should aim to make it use a more hook-like flow, such that if the decision to move it to an extension is made it can be done.

My mental image is that there would be a class like CRM_Contribute_DefferredRevenue() and function snippets like

```
 // CRM-16189
 +    CRM_Core_BAO_FinancialTrxn::generateRevenueRecognitionDate($params, $contributionID);
      if ($contributionID && !empty($params['revenue_recognition_date']) && !empty($params['prevContribution'])
        && !($contributionStatus[$params['prevContribution']->contribution_status_id] == 'Pending')
        && !self::allowUpdateRevenueRecognitionDate($contributionID)
.....

```

would sit on that class eg 

```
function hookPreContribution() {
 // CRM-16189
 +    CRM_Core_BAO_FinancialTrxn::generateRevenueRecognitionDate($params, $contributionID);
      if ($contributionID && !empty($params['revenue_recognition_date']) && !empty($params['prevContribution'])
        && !($contributionStatus[$params['prevContribution']->contribution_status_id] == 'Pending')
        && !self::allowUpdateRevenueRecognitionDate($contributionID)
}
```

with the Contribution BAO having just before the pre Hook 


CRM_Contribute_DefferredRevenue::hookPreContribution(....);

File: CRM/Contribute/BAO/Query.php (hunk 1 above)
Comment (JoeMurray) [inline CRM/Contribute/BAO/Query.php]: We're going to leave this refactoring as out of scope.

File: CRM/Core/BAO/FinancialTrxn.php [hunk 3]
```diff
@@ -555,34 +555,72 @@ public static function getPartialPaymentTrxn($contribution, $params) {
    *
```
Comment (JoeMurray) [inline CRM/Core/BAO/FinancialTrxn.php]: Please add to description that the membership revenue amount returned is going to be the amount to be recognized as of the $receiveDate for accrual accounting purposes

File: CRM/Core/BAO/FinancialTrxn.php [hunk 4]
```diff
@@ -555,34 +555,72 @@ public static function getPartialPaymentTrxn($contribution, $params) {
...
    * @param array $lineItem
    *
+   * @param date $receiveDate
+   *
    * @return array
    */
-  public static function getMembershipRevenueAmount($lineItem) {
+  public static function getMembershipRevenueAmount($lineItem, $receiveDate) {
```
Comment (JoeMurray) [inline CRM/Core/BAO/FinancialTrxn.php]: getMembershipAmountToRecognize would be a better name.

File: CRM/Core/BAO/FinancialTrxn.php [hunk 5]
```diff
@@ -591,37 +629,50 @@ public static function getMembershipRevenueAmount($lineItem) {
    * Create transaction for deferred revenue.
```
Comment (JoeMurray) [inline CRM/Core/BAO/FinancialTrxn.php]: Change to 'Create db entries to recognize deferred revenue appropriately in the future'

File: CRM/Core/BAO/FinancialTrxn.php [hunk 6]
```diff
@@ -591,37 +629,50 @@ public static function getMembershipRevenueAmount($lineItem) {
...
+   * @param int|CRM_Contribute_BAO_Contribution $contribution
    * @param bool $update
-   *
    * @param string $context
    *
+   * @return void
    */
-  public static function createDeferredTrxn($lineItems, $contributionDetails, $update = FALSE, $context = NULL) {
-    if (empty($lineItems)) {
+  public static function createDeferredTrxn($lineItems, $contribution, $update = FALSE, $context = NULL) {
```
Comment (JoeMurray) [inline CRM/Core/BAO/FinancialTrxn.php]: Please revert the change from $contributionDetails to $contribution. Make the calling context look up the contribution from the contribution id if necessary.

File: CRM/Price/BAO/LineItem.php (hunk 2 above)
Comment (JoeMurray) [inline CRM/Price/BAO/LineItem.php]: Agreed. Drop this extraneous comment.

File: CRM/Core/BAO/FinancialTrxn.php [hunk 7]
```diff
@@ -591,37 +629,50 @@ public static function getMembershipRevenueAmount($lineItem) {
...
+
+    $revenueRecognitionDate = CRM_Utils_Array::value('revenue_recognition_date', $contributionDetails);
     if (!CRM_Utils_System::isNull($revenueRecognitionDate)) {
       $statuses = CRM_Contribute_PseudoConstant::contributionStatus(NULL, 'name');
       if (!$update
-        && (CRM_Utils_Array::value($contributionDetails->contribution_status_id, $statuses) != 'Completed'
-          || (CRM_Utils_Array::value($contributionDetails->contribution_status_id, $statuses) != 'Pending'
-            && $contributionDetails->is_pay_later)
-          )
+        && (!(CRM_Utils_Array::value($contributionDetails['contribution_status_id'], $statuses) == 'Completed'
```
Comment (JoeMurray) [inline CRM/Core/BAO/FinancialTrxn.php]: Let's move the code cleanup into a separate commit for this PR

File: CRM/Core/BAO/FinancialTrxn.php [hunk 8]
```diff
@@ -668,10 +719,10 @@ public static function createDeferredTrxn($lineItems, $contributionDetails, $upd
         }
         foreach ($results['values'] as $result) {
           if ($result['account_relationship'] == $accountRel) {
-            $trxnParams['to_financial_account_id'] = $result['financial_account_id'];
+            $trxnParams['from_financial_account_id'] = $result['financial_account_id'];
```
Comment (JoeMurray) [inline CRM/Core/BAO/FinancialTrxn.php]: Let's move this and L725 into a separate PR for CRM-20585.

File: CRM/Event/Form/Participant.php [hunk 9]
```diff
@@ -958,7 +958,7 @@ public function postProcess() {
       $additionalParticipantDetails = array();
       if (CRM_Contribute_BAO_Contribution::checkContributeSettings('deferred_revenue_enabled')) {
         $eventStartDate = CRM_Core_DAO::getFieldValue('CRM_Event_DAO_Event', $this->_eventId, 'start_date');
-        if ($eventStartDate) {
+        if (strtotime($eventStartDate) > strtotime(date('Ymt'))) {
```
Comment (JoeMurray) [inline CRM/Event/Form/Participant.php]: previously merged to core in a different PR

File: CRM/Event/Form/Registration/Confirm.php [hunk 10]
```diff
@@ -1027,7 +1027,7 @@ public static function processContribution(
           $form->_values
         )
       );
-      if ($eventStartDate) {
+      if (strtotime($eventStartDate) > strtotime(date('Ymt'))) {
```
Comment (JoeMurray) [inline CRM/Event/Form/Registration/Confirm.php]: previously merged to core in a separate PR - please remove from here.

File: tests/phpunit/CRM/Core/BAO/FinancialTrxnTest.php [hunk 11]
```diff
@@ -135,6 +135,7 @@ public function testGetPartialPaymentTrxn() {
   public function testCreateDeferredTrxn() {
     Civi::settings()->set('contribution_invoice_settings', array('deferred_revenue_enabled' => '1'));
     $cid = $this->individualCreate();
+    $financialTypeId = 4;
```
Comment (JoeMurray) [inline tests/phpunit/CRM/Core/BAO/FinancialTrxnTest.php]: putting in variable since reusing at L155 and L167

**Comment by eileenmcnaughton** [comment]
@pradpnayak @JoeMurray can we close this? I feel it is superceded

**Comment by eileenmcnaughton** [comment]
@pradpnayak @JoeMurray per previous comment I'm going to close this & let you open one or more appropriate PRs for any parts that are valid
